### PR TITLE
Added JAX pytrees notebook

### PR DESCRIPTION
**Pull request by gnecula**
Also added docstrings to the tree_util module.

### CODE REVIEW

File: docs/notebooks/JAX_pytrees.ipynb [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "sgUJpiXSsRSi",
+        "colab_type": "text"
+      },
+      "source": [
+        "Pytrees containers can be lists, tuples, dicts, namedtuple. Leaves as numeric and ndarrays:"
```
Comment (shoyer) [inline docs/notebooks/JAX_pytrees.ipynb]: minor language suggestion:
```suggestion
        "Pytrees containers can be lists, tuples, dicts, namedtuple. Numeric and ndarray values are treated as leaves:"
```

**Review by mattjj — APPROVED** [review]
LGTM! Awesome, this is really clean and thorough :)

File: docs/notebooks/JAX_pytrees.ipynb [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "sgUJpiXSsRSi",
+        "colab_type": "text"
+      },
+      "source": [
+        "Pytrees containers can be lists, tuples, dicts, namedtuple. Numeric and ndarray values are treated as leaves:"
```
Comment (hawkinsp) [inline docs/notebooks/JAX_pytrees.ipynb]: The comment about numeric and ndarray values is a bit misleading (indeed, see the following section.)

Nit: `NamedTuple`. We also support a few more things via custom registration (e.g., `OrderedDict`).

`None` is also currently considered an interior node.

**Review by hawkinsp — APPROVED** [review]

File: docs/notebooks/JAX_pytrees.ipynb [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+        "value_flat, value_tree = tree_flatten(value_structured)\n",
+        "print(\"value_flat={}\\nvalue_tree={}\".format(value_flat, value_tree))\n",
+        "\n",
+        "# Transform the flt value list using an element-wise numeric transformer\n",
+        "transformed_flat = list(map(lambda v: v * 2., value_flat))\n",
+        "print(\"transformed_flat={}\".format(transformed_flat))\n",
+        "\n",
+        "# Reconstruct the structured output, using the original \n",
+        "transformed_structured = tree_unflatten(value_tree, transformed_flat)\n",
+        "print(\"transformed_structured={}\", transformed_structured)"
```
[the remark this answers is not in the document]
Comment (gnecula) [reply docs/notebooks/JAX_pytrees.ipynb]: Yes, thanks

File: docs/notebooks/JAX_pytrees.ipynb (hunk 2 above)
Comment (gnecula) [reply docs/notebooks/JAX_pytrees.ipynb]: I have added OrderedDict and None to the list of internal nodes. I changed the comment about the numeric values:

"Other types of values, including numeric and ndarray values, are treated as leaves"

I hope that's what you meant.

**Pull request merged**